### PR TITLE
Implement musti-signing signature verification.

### PR DESCRIPTION
**Pull request by dzmitryhil**
# Description

The PR contains logic and tests for the multisig-signing transactions signature verification. 

The changes in the `data/encoder.go` are required for the signature verification, since with the current logic the `ignoreSigningFields` works incorrectly with the slices/arrays. If we ignore the array, for example `Signers`, it adds the start and end of the array to the decoded transaction, without  the array content, which is invalid, and leads to the signature verification failure.

### CODE REVIEW

**Comment by donovanhide** [comment]
Looks good, but a little bit confused by your reference to arrays, which are fixed length in Go. Are you talking about the ripple binary format ST_ARRAY? Probably should be explicit in your commit message if so. There is no ST_SLICE :-)

**Comment by dzmitryhil** [comment]
> Looks good, but a little bit confused by your reference to arrays, which are fixed length in Go. Are you talking about the ripple binary format ST_ARRAY? Probably should be explicit in your commit message if so. There is no ST_SLICE :-)

Sure, updated the commit message.

**Comment by dzmitryhil** [comment]
> > Looks good, but a little bit confused by your reference to arrays, which are fixed length in Go. Are you talking about the ripple binary format ST_ARRAY? Probably should be explicit in your commit message if so. There is no ST_SLICE :-)
> 
> Sure, updated the commit message.

@donovanhide Is it good now?

**Comment by donovanhide** [comment]
Sorry, not sure if your force-push worked. Can't see any changes to the commit message.

**Comment by dzmitryhil** [comment]
@donovanhide Check it [here](https://github.com/rubblelabs/ripple/pull/85/commits/214e2769cb13c0a1fb3569593eaeab36b4de0c12).

File: data/signing_test.go [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+	// important for the multi-signing
+	tx.TxBase.SigningPubKey = &PublicKey{}
+	return &tx
+}
+
+func genSeed(t *testing.T) *Seed {
+	const letterBytes = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
+	password := make([]byte, 20)
+	for i := range password {
+		password[i] = letterBytes[rand.Intn(len(letterBytes))]
```
Comment (donovanhide) [inline data/signing_test.go]: This is obviously a very bad way to generate a password :-) I know it's only a test, but people copy and paste code. Maybe either just pass the password as a fixed string for each signer to genSeed, or use: 

https://pkg.go.dev/crypto/rand#Read

and convert the []byte to a string.

Comment (dzmitryhil) [reply data/signing_test.go]: Done

File: data/signing_test.go [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+	tx.TxBase.SigningPubKey = &PublicKey{}
+	return &tx
+}
+
+func genSeedFromPassword(t *testing.T, password string) *Seed {
+	seedFromPass, err := crypto.GenerateFamilySeed(password)
+	if err != nil {
+		t.Fatal(err)
+	}
+	seed, err := NewSeedFromAddress(seedFromPass.String())
```
Comment (donovanhide) [inline data/signing_test.go:114]: You don't need this second seed. Use the Payload() function.

https://github.com/rubblelabs/ripple/blob/master/crypto/key_test.go#L65

File: data/signing_test.go [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+	}
+	if len(invalidSigners) != 0 {
+		t.Fatal(err)
+	}
+
+	// update one signature
+	tx.Signers[0].Signer.TxnSignature = tx.Signers[1].Signer.TxnSignature
+	valid, invalidSigners, err = CheckMultiSignature(tx)
+	if valid {
+		t.Fatal(err)
```
Comment (donovanhide) [inline data/signing_test.go]: Why is this same err in all the Fatal calls? You need an appropriate error string for each test failure. The first test should be `if err!=nil`

Comment (dzmitryhil) [reply data/signing_test.go]: You are right, updated.

File: data/signing_test.go (hunk 2 above)
Comment (dzmitryhil) [reply data/signing_test.go:114]: Could you elaborate on it a bit?
Becase in the example you sent `NewECDSAKey(seed.Payload())` the key is generated from the seed, and in the test we generate the Key from the seed `seed1.Key(ECDSA)` which is similar.

Comment (donovanhide) [reply data/signing_test.go:114]: I guess it's just a bit of pointless backwards and forwards between unnecessary base58 encoding. I wrote all this code 9 years ago and can barely remember half of it!
I don't really have time to do full code reviews on this and other PRs, paying work takes priority :-) I'd just say read all the existing tests, make sure all your tests log the reason for failure with a contextual sentence, expected and result text and use testing.FatalF to do so:

https://pkg.go.dev/testing#T.Fatalf

Happy to merge once done.

If you want a challenge, you could port all the existing tests from go.check to testing.T 👍

Comment (dzmitryhil) [reply data/signing_test.go:114]: I've added `testing.FatalF` where it gives more information. 
What is regarding to `go.check `, I might do it in the future, once start touching the code it tests.

**Comment by dzmitryhil** [comment]
@donovanhide Do you want me to change anything else in that PR ?

**Comment by donovanhide** [comment]
I left comments here:
https://github.com/rubblelabs/ripple/commit/18baf659b57cbedf6adfbe2d62de3de2e6a7f344#comments
I think force-pushing fixes to comments might be disrupting the PR review process... 

I don't really have time to review all this and your colleague's work. The changes you are asking for have no effect on my private usage of this library. Might be better for you to fork this library and just do whatever it is you need for your work. I don't really want to deal with other users complaining about broken API's and bugs, there is no upside :-) This library is ancient and not really what I would write today.

**Comment by dzmitryhil** [comment]
> I left comments here: [18baf65#comments](https://github.com/rubblelabs/ripple/commit/18baf659b57cbedf6adfbe2d62de3de2e6a7f344#comments) I think force-pushing fixes to comments might be disrupting the PR review process...
> 
> I don't really have time to review all this and your colleague's work. The changes you are asking for have no effect on my private usage of this library. Might be better for you to fork this library and just do whatever it is you need for your work. I don't really want to deal with other users complaining about broken API's and bugs, there is no upside :-) This library is ancient and not really what I would write today.

No problem, I'll cancel that PR and create to our fork, just wanted them to be on-track.

**Pull request closed without merging**